java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java

```diff
@@ -524,34 +525,50 @@ public void visitLambdaExpression(LambdaExpressionTree tree) {
   }
 
   private void refineLambdaType(LambdaExpressionTreeImpl lambdaExpressionTree, JavaType lambdaType) {
-    JavaType samReturnType = (JavaType) getSamMethod(lambdaType).returnType().type();
+    JavaSymbol.MethodJavaSymbol samMethod = getSamMethod(lambdaType);
+    if(samMethod == null) {
+      return;
+    }
+    JavaType samReturnType = (JavaType) samMethod.returnType().type();
     JavaType capturedReturnType = resolve.resolveTypeSubstitution(samReturnType, lambdaType);
     if (lambdaExpressionTree.body().is(Tree.Kind.BLOCK) || capturedReturnType.is("void")) {
       // TODO compute return type for lambda with block and return statements
       return;
     }
     JavaType refinedReturnType = (JavaType) ((AbstractTypedTree) lambdaExpressionTree.body()).symbolType();
-    if (refinedReturnType != capturedReturnType && lambdaType.isParameterized()) {
+    refineType(lambdaExpressionTree, lambdaType, capturedReturnType, refinedReturnType);
+  }
+
+  private void refineType(AbstractTypedTree expression, JavaType expressionType, JavaType capturedReturnType, JavaType refinedReturnType) {
+    if (refinedReturnType != capturedReturnType && expressionType.isParameterized()) {
```

Comment:
Statement nesting can be avoided by inverting this `if` and returning early from the method.